libtransact/src/scheduler/multi/shared.rs

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2019 Cargill Incorporated
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -----------------------------------------------------------------------------
+ */
+
+//! Internal MultiScheduler state shared across threads.
+
+use crate::protocol::batch::BatchPair;
+use crate::scheduler::{BatchExecutionResult, Scheduler};
+
+use std::collections::{HashMap, HashSet};
+
+/// Stores all MultiScheduler data which is shared between threads.
+pub struct MultiSchedulerShared {
+    finalized: bool,
+    /// The MultiScheduler's result callback; this is called when all sub-schedulers have returned
+    /// the same result for a batch.
+    result_callback: Option<Box<Fn(Option<BatchExecutionResult>) + Send>>,
+    /// Tracks which sub-schedulers have returned matching results for the given batch pair.
+    pending_results: HashMap<BatchPair, (Option<BatchExecutionResult>, HashSet<usize>)>,
+    /// The sub-schedulers of this MultiScheduler.
+    schedulers: Vec<Box<dyn Scheduler + Send>>,
+}
+
+impl MultiSchedulerShared {
+    pub fn new(schedulers: Vec<Box<dyn Scheduler + Send>>) -> Self {
+        MultiSchedulerShared {
+            finalized: false,
+            result_callback: None,
+            pending_results: HashMap::new(),
+            schedulers,
+        }
+    }
+
+    pub fn finalized(&self) -> bool {
+        self.finalized
+    }
+
+    pub fn result_callback(&self) -> &Option<Box<Fn(Option<BatchExecutionResult>) + Send>> {
+        &self.result_callback
+    }
+
+    pub fn set_result_callback(&mut self, callback: Box<Fn(Option<BatchExecutionResult>) + Send>) {
+        self.result_callback = Some(callback);
+    }
+
+    pub fn add_batch(&mut self, batch: BatchPair) {
+        self.schedulers.iter_mut().for_each(|scheduler| {
+            scheduler.add_batch(batch.clone());
+        });
+        self.pending_results.insert(batch, (None, HashSet::new()));
```

Comment:
How should duplicate batches be handled? If a duplicate batch is added, the pending result for that batch will be overridden here.

Reply:
I think adding duplicate batches to the scheduler should be considered a programming error on behalf of the calling code (the code using the scheduler).

Reply:
Are you suggesting we should panic here? Or just ignore this possibility?

Reply:
Or possibly add_batch() in the scheduler needs to return a result so we should propagate up the error.

Reply:
I'm not sure returning a result from add_batch is particularly interesting for schedulers in general; the serial and parallel schedulers probably don't care if they get duplicate batches, they'll just execute the transactions again when they get around to them.

Reply:
It does matter for serial and parallel schedulers, because you wouldn't be able to differentiate the result with the correct one. It might actually be problematic inside the parallel scheduler for other reasons as well, but I think this first reason is sufficient to change the API and make sure we don't allow ingress of duplicate batches. Or more specifically, duplicate batches which are in-progress. (The scheduler shouldn't keep track of batches long-term for this purpose, so it would bascially just ensure all the batches currently known to the scheduler are unique.)

Reply:
Added a commit that updates add_batch to return a Result. Creates a new SchedulerError enum. Also added a SchedulerFinalized variant that's used instead of panicking when the add_batch method is called on a finalized scheduler.

Reply:
This update was moved to #85